parser/NodeCopying.cc

```diff
@@ -0,0 +1,353 @@
+        [&](const parser::Self *self) { result = std::make_unique<Self>(self->loc); },
+        [&](const parser::Send *send) {
+            result = std::make_unique<Send>(send->loc, deepCopy(send->receiver.get()), send->method, send->methodLoc,
+                                            deepCopyVec(send->args));
+        },
+        [&](const parser::Shadowarg *shadowarg) {
+            result = std::make_unique<Shadowarg>(shadowarg->loc, shadowarg->name);
+        },
+        [&](const parser::Splat *splat) { result = std::make_unique<Splat>(splat->loc, deepCopy(splat->var.get())); },
+        [&](const parser::SplatLhs *splatLhs) {
+            result = std::make_unique<SplatLhs>(splatLhs->loc, deepCopy(splatLhs->var.get()));
+        },
+        [&](const parser::String *string) { result = std::make_unique<String>(string->loc, string->val); },
+        [&](const parser::Super *super) { result = std::make_unique<Super>(super->loc, deepCopyVec(super->args)); },
+        [&](const parser::Symbol *symbol) { result = std::make_unique<Symbol>(symbol->loc, symbol->val); },
+        [&](const parser::True *true_) { result = std::make_unique<True>(true_->loc); },
+        [&](const parser::Undef *undef) { result = std::make_unique<Undef>(undef->loc, deepCopyVec(undef->exprs)); },
+        [&](const parser::UnlessGuard *unlessGuard) {
+            result = std::make_unique<UnlessGuard>(unlessGuard->loc, deepCopy(unlessGuard->condition.get()));
+        },
+        [&](const parser::Until *until) {
+            result = std::make_unique<Until>(until->loc, deepCopy(until->cond.get()), deepCopy(until->body.get()));
+        },
+        [&](const parser::UntilPost *untilPost) {
+            result = std::make_unique<UntilPost>(untilPost->loc, deepCopy(untilPost->cond.get()),
+                                                 deepCopy(untilPost->body.get()));
+        },
+        [&](const parser::When *when) {
+            result = std::make_unique<When>(when->loc, deepCopyVec(when->patterns), deepCopy(when->body.get()));
+        },
+        [&](const parser::While *while_) {
+            result = std::make_unique<While>(while_->loc, deepCopy(while_->cond.get()), deepCopy(while_->body.get()));
+        },
+        [&](const parser::WhilePost *whilePost) {
+            result = std::make_unique<WhilePost>(whilePost->loc, deepCopy(whilePost->cond.get()),
+                                                 deepCopy(whilePost->body.get()));
+        },
+        [&](const parser::XString *xString) {
+            result = std::make_unique<XString>(xString->loc, deepCopyVec(xString->nodes));
+        },
+        [&](const parser::Yield *yield) { result = std::make_unique<Yield>(yield->loc, deepCopyVec(yield->exprs)); },
+        [&](const parser::ZSuper *zSuper) { result = std::make_unique<ZSuper>(zSuper->loc); },
+        [&](const parser::Node *other) {
+            ENFORCE(false, "Unimplemented node type: {}", ((parser::Node *)other)->nodeName());
```

Comment:
This seems fishy but I couldn't access `nodeName` otherwise, open to suggestions.

Reply:
https://github.com/sorbet/sorbet/commit/e3be9f76912cdfd39ba21302c8332d5cac4c2724

(I can't figure out how to push this to your PR, sorry.)

Reply:
#8820